src/AI/SpenderAI.java

```diff
@@ -0,0 +1,325 @@
+      }
+      else if( unitQueue.isEmpty() )
+      {
+        stateChange = false; // There's been no gamestate change since we last iterated through all the units, since we're about to do just that
+        for( Unit unit : myCo.units )
+        {
+          if( unit.isTurnOver )
+            continue; // No actions for stale units.
+          unitQueue.offer(unit);
+        }
+      }
+
+      Queue<Unit> travelQueue = new ArrayDeque<Unit>();
+      // Handle actions for each unit the CO owns.
+      while (!unitQueue.isEmpty())
+      {
+        Unit unit = unitQueue.poll();
+        boolean foundAction = false;
+
+        // Find the possible destinations.
+        ArrayList<XYCoord> destinations = Utils.findPossibleDestinations(unit, gameMap);
+
+        for( XYCoord coord : destinations )
+        {
+          // Figure out how to get here.
+          Path movePath = Utils.findShortestPath(unit, coord, gameMap);
+
+          // Figure out what I can do here.
+          ArrayList<GameActionSet> actionSets = unit.getPossibleActions(gameMap, movePath);
+          for( GameActionSet actionSet : actionSets )
+          {
+            // See if we have the option to attack.
+            if( actionSet.getSelected().getType() == GameAction.ActionType.ATTACK )
+            {
+              actions.offer(actionSet.getSelected());
+              foundAction = true;
+              break;
+            }
+
+            // Otherwise, see if we have the option to capture.
+            if( actionSet.getSelected().getType() == GameAction.ActionType.CAPTURE )
+            {
+              actions.offer(actionSet.getSelected());
+              capturingProperties.add(coord);
+              foundAction = true;
+              break;
+            }
+          }
+          if( foundAction )
+            break; // Only allow one action per unit.
+        }
+        if( foundAction )
+        {
+          stateChange = true;
+          break; // Only one action per getNextAction() call, to avoid overlap.
+        }
+        else
+        {
+          travelQueue.offer(unit); // if we can't do anything useful right now, consider just moving towards a useful destination
+        }
+      }
+
+      Queue<Unit> waitQueue = new ArrayDeque<Unit>();
+
+      // If no attack/capture actions are available now, just move towards a non-allied building.
+      if( actions.isEmpty() && !stateChange )
+      {
+        while (!travelQueue.isEmpty())
+        {
+          Unit unit = travelQueue.poll();
+
+          // Find the possible destinations.
+          ArrayList<XYCoord> destinations = Utils.findPossibleDestinations(unit, gameMap);
+
+          Utils.sortLocationsByDistance(new XYCoord(unit.x, unit.y), unownedProperties);
+          if( !unownedProperties.isEmpty() ) // Sanity check - it shouldn't be, unless this function is called after we win.
+          {
+            log(String.format("  Seeking a property to send %s after", unit.toStringWithLocation()));
```

Comment:
I feel like this should be limited to units with the ability to capture. Maybe send non-capturing units to a) the nearest enemy, or b) the enemy HQ.
Just for grins, I made local hacks to only consider enemy HQs for units that can't capture, and it shuts down ISAI that much faster since it goes straight for the jugular.

Reply:
That would objectively improve the AI, yes, but having artillery sit on enemy cities with a halo of other units orbiting them has a certain charm, heh

Reply:
For the record, my intent with this AI was just to get it to use its resources more fully.
Being *smart* with them is a subject for another pull request.

Reply:
I like the idea of having a number of AIs to choose from with variable weaknesses and... okay, just variable weaknesses is fine.
You could also note that it almost makes a point of rushing Artillery to the front line, but I'm not sure that's worth fixing here either.

The problem with fixing this particular problem is I don't know what I would want it to do instead.
You could have it go for the HQ, but then it wouldn't move off of *that* either.
You could have it go straight for the enemies, but that might be undesirable in other ways (splitting forces, running headlong into danger).

Reply:
I probably wouldn't worry about Artillery placement, since that can be a tricky beast, and though the current behavior is suboptimal, it isn't flat wrong per se... just amateurish.
I think going straight for the HQ is fine, because if they reach it, everyone defending it is probably dead anyway. The deadlock condition for HQs is not as troublesome as for other buildings, and even if it is, I'd rather deadlock on one building than all of them.
Going straight for enemies is probably also fine - no it's not optimal, but it's at least understandable in a way that preventing yourself from capturing a city isn't.

Reply:
To be clear, I'm not asking you to make an optimal AI before merging - I just want one that doesn't grotesquely shoot itself in the foot (unless that's like, a feature, man).

Reply:
It's a feature *now*.
But okay, I'll fix it.